### ISSUE TITLE
Setting to require alt text for image uploads.

### ISSUE THREAD

**Issue by rkingett** [mastodon/mastodon#13894]
### Pitch

I propose that a setting be added for instance hosts to require image descriptions, alt text, from their users  upon upload, as well as user settings to where they can toggle if they want to restrict image uploads without alt text.

Instance hosts could require image descriptions when uploading an image by way of alt text.

Users can set it so they are forced to add alt text for the images they upload.

### Motivation

Even though Mastodon makes it very easy for people to add alt text, so few people do it.

The app, Toot, has this feature. I'd like it to be a platform wide setting for hosts and users.

**Comment by SuperSandro2000** [comment]
I don't think this is a good idea cause people will type in garbage if they a re forced to do so.

**Comment by Cassolotl** [comment]
I think discussions have come up about this before. I saw someone who uses a screenreader give a mini-tutorial about image descriptions a while back and they basically said that unnecessary words is just a waste of time and attention, and that includes image descriptions where it's clear from context what the image is. Like, "Check out this adorable puppy I saw at the shops today!" and then a photo of a dog, the photo doesn't need an image description at all.

Someone could get around compulsory image descriptions by just putting a dot in the image description box, and I don't know if that would be inconvenient enough for people with screenreaders that they'd prefer to just have image descriptions optional - can anyone who uses a screenreader chip in?

**Comment by rkingett** [comment]
Good points. Could we have a setting for users, not instance administrators, to just hide all images without alt text completely on a user level, not an admin level?

But if the above setting was never turned on by an instance administrator, nobody would have to do it.

How about the following?

* Add a user setting they can toggle to require alt text when they upload pictures.
* Have a setting where users can completely hide images without alt text in all timelines.

**Comment by Sasha-Sorokin** [comment]
Why would you even want to hide all images without alternative text? This will only hurt users in the future¹ because they will be unaware about the context, that there is image attached to the post. Hiding posts with images with no alternative text is also harmful because it can simply empty timelines of many posts that have optional images.

Neither you should force users into filling out the descriptions, because they will then skip this step by typing out the garbage as pointed by people above and we already have enough garbage alternative texts, thanks to Pleroma that puts filenames in there.

Instead, what I think we can and should do, is to train and encourage users to fill alternative texts using, say, coachmark and linking the tutorial:

<p align="center">
<img src="https://user-images.githubusercontent.com/10401817/83452404-d5ecef80-a482-11ea-9f8c-2c0b720b90d9.png" alt="Screenshot of popup near Edit button for media attachment, suggesting user to learn more about alternative descriptions and add one">
</p>
<p align="center">
<i>Concept image of idea I had for a long time.</i>
</p>


Still optional, but may motivate more people to fill alternative texts. This idea however best be tracked in the other issue, I think.

***1** — currently there is an unfixed issue where images without alternative texts don't have `alt` attribute at all, making these be skipped by screenreaders. The fix will be to have simple “Image” alternative text to make SR users aware of it.*

**Comment by rkingett** [comment]
I like your idea a lot. Could you link that issue?

I just got tired of seeing images without alt text so wanted a quick way to get rid of all of them when I felt like it. I wanted it more like a toggle, that users can toggle on and off. I know many screen reader users, apart from me, would willingly filter out inaccessible content, given the chance, important or not. I know I'd do it in a minute, if I could.

If I had the chance, I'd willingly do it. I know others would too. But, I really like your idea also. Could you link the other issue?

**Comment by Kjwon15** [comment]
Why not just force to open edit description dialog after uploading an image?

**Comment by byronalley** [comment]
Seems like the ideal UI behaviour would guide people to the right choice without imposing anything.  Right now the description dialog is hard to see (via an "edit" link) and it's not clear that "edit" means "give it a description" rather than "edit this image" (eg. enhance, change contrast, shrink?).

The more obvious choice would be to have the "describe image" field appear automatically once you've added an image.

Then it could have text, eg: "Unless this image is already described by the text in your post, it should have a description for users with screen readers.

- If you haven't entered any text, the confirm button changed to something like [Post Without Description]
- Once you enter a description, button switches to [Post]

But already, just removing a step for the alt text would likely increase the use of it.

**Comment by rkingett** [comment]
I approve with  @byronalley completely.

**Comment by notverymoe** [comment]
This is something I'm interested in as the current "no description" warning isn't aggressive enough for me, I often jump the gun as soon as I hit the end of my post (ADHD in my case, so it'd be a great accessibility option imo). In particular I'd like to see a user option to enable a warning dialog when you attempt to post images without alt-text, this warning would be dismissible for the mentioned case where alt-text doesn't make sense

I also think  @byronalley 's idea is good too and that both could be implemented.

**Comment by rkingett** [comment]
People think this would be enabled for every user, which, it wouldn't. What I'd like is to give users the ability to enable this requirement on their own accounts without it effecting other users.

Many users enable the setting in the iOS app, Toot. If this is a user setting, you don't need to turn it on.

I'd like to hear how helpful [this new reminder is to users[(https://github.com/mastodon/mastodon/pull/17845) as well.

**Comment by mgifford** [comment]
Ideally it is harder to upload an image without alt text, as it is to provide the alt text. 

At the very least, users should be alerted and told that adding alt text will help with both accessibility and findability within Mastodon. 

Users may want to have the option of requiring alt text, or for that matter skipping the reminder. The default should be one that encourages users to produce more accessible content. 

Mastodon is a publishing tool, so should be thinking about https://www.w3.org/TR/ATAG20/ which outlines that the authoring interface should be accessible, and should help authors create accessible content.

**Comment by byronalley** [comment]
It's important to remember that _bad_ alt text is typically worse than no alt text at all. So that's the tricky thing here: you want to make it easy for people who want to do a good job of it, but you don't want a new Mastodon user to see it, think "ugh I don't get what this is, why do I need it?" and type "ughghghghgh" or "here is some alt text" or "DFSDQQZR_05324516432.JPEG" or "an image I uploaded." That's why pushing too hard for alt text will backfire.

**Comment by mgifford** [comment]
I'm fine with that @rkingett just as long as it is enabled by default. It is ultimately up to the author most of the time. Some instances may want to force their users to use alt text, but not sure how many. 

In Drupal we decided that images uploaded in the CKEditor were by default meaningful (rather than decorative) images. It is a good assumption for Mastodon too I think.

**Comment by rkingett** [comment]
This is mostly solved by [this Pull request, here](https://github.com/mastodon/mastodon/pull/33760)
